### PR TITLE
Restructure 07-contributor-guide

### LINKED ISSUE #106: [BUG] workflow doc doesn't follow the rebase workflow we want to use
**Describe the bug**
We've been merging main into feature branches and this results in a very messy commit history - we are going to move to rebasing instead

**To Reproduce**
Update your feature branch
Update main

**Expected behavior**
The user rebases the changes from main into their branch and does not use pull and merge

### PR DESCRIPTION
**Pull request by kellijohnson-NOAA**
Sorry @Bai-Li-NOAA I didn't mean to hit create PR before editing this. This is a LARGE change to 07-contributor-guide. I am fine if there are parts that you would like me to rethink. Happy to make changes here. I just wanted to get it on your radar that I had put some time into this. Which all started as wanting to close #106 and working on #144.

I still have a few more things that I would like to do (see list below). I am not sure if I should make this a draft request and do them or if they can be done later as a separate PR.
TODO:
* Find where the review guidance exists online for both sections and reference the links
* document makevars and makevars.win
* document distributions folder
* create GitHub action to open an issue in collaborative_workflow when a new folder is added to NOAA-FIMS/FIMS/inst/include. Kathryn is going to help but I need to find how to know that a new folder was added, i.e., what is the reference to compare to?

# What is the feature?
* Make the contributor guide more relevant to us now and our future selves.

# How have you implemented the solution?
The chapter was reconfigured to start with style guides, go into github and issues, then github PR, then talk about why we use git.

# Does the PR impact any other area of the project?

### CODE REVIEW

**Comment by Bai-Li-NOAA** [comment]
@kellijohnson-NOAA. Thank you so much for updating the contributor guide! I can review the pull request early next week. 

Regarding the do-do items, they seem to be new additions to the book or new features. Since they don't seem closely related to issues #106 and #144, feel free to address them in a separate pull request at a later time, unless you plan to work on them before next week.

File: 07-contributor-guide.Rmd
Comment (Bai-Li-NOAA) [inline 07-contributor-guide.Rmd:41]: Do we want to document the FIMS file structure here in the collaborative workflow? It might be easier to remember to maintain the descriptions if we keep them within the FIMS repo. We can create short README.md files in the distributions, interface, and other folders to provide brief introductions (see [googletest example](https://github.com/google/googletest/tree/main/googlemock)). However, I'm not sure if these README.md files follow good practices for developing R packages.

File: 07-contributor-guide.Rmd
Comment (Bai-Li-NOAA) [inline 07-contributor-guide.Rmd:112]: Typo in the number? Should 1 be changed to 2, or are we relying on Markdown to automatically index the list?

File: 07-contributor-guide.Rmd
Comment (Bai-Li-NOAA) [inline 07-contributor-guide.Rmd:178]: I wonder if we need to provide additional guidance in the book, Maybe we can just provide a link to the [Checklist in the FIMS](https://github.com/NOAA-FIMS/FIMS/blob/main/.github/workflows/pr-checklist.yml#L31C12-L41C74) repo to avoid updating the list in two places.

File: 07-contributor-guide.Rmd
Comment (Bai-Li-NOAA) [inline 07-contributor-guide.Rmd:282]: Do we want to recommend squashing commits before pushing final changes to GitHub?

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:112]: I am assuming that markdown will index the list.

**Review by Bai-Li-NOAA — APPROVED** [review]
This is a lot of work. Thanks, @kellijohnson-NOAA, for revising Chapter 7. I only had a few minor, non-blocking suggestions. 

I think this PR could potentially close issue #107, as you've added a commit messages section in the book, and we have some instructions in the FIMS `pr-checklist.yml` regarding nitpicking. 

I also noticed that I am the assignee for this PR. Would you like me to work on the TODO items in the PR?

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:178]: This is a great idea @Bai-Li-NOAA. Though, I think that I will just link to the .yml file itself rather than specific lines because the whole file is relevant and I would hate to see the link to specific lines break if we were to edit that file. Does that sound okay?

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:282]: I will add some language about squashing but just as additional guidance, i.e., not a requirement.

Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:282]:
```
4. Consider interactively rebasing your commits to squash similar commits
into a single commit or rearrange the order of your commits. The following
example is for the situation where you want to look at the last ten commits.
{r eval = FALSE}
$ git rebase -i head~10
```

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:41]: I am open to ideas here. I talked to @k-doering-NOAA about potentially creating a GitHub action that creates an issue in the collaborative_workflow repo when folders are added to the FIMS file structure. But that might be overly complicated. @Bai-Li-NOAA what do you think would be the best for training purposes? If we had documentation in the folders themselves that we could "scrape" when making presentations or training material I think that would be ideal. I know that we can add files to .Rbuildignore so they are ignored when building FIMS if we want READMEs in the directories themselves.

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:178]: @Bai-Li-NOAA how much of this section should I delete? Should all bulleted points below line 178 be deleted and the last sentence of line 178?

File: 07-contributor-guide.Rmd
Comment (Bai-Li-NOAA) [reply 07-contributor-guide.Rmd:41]: I like the idea of automatically updating training materials to introduce the FIMS file structure. We could include short README.md files in each folder and use a vignette to automatically compile all *.md files from inst/include. Additionally, we might consider using Doxygen to generate a [file dependencies diagram](https://www.doxygen.nl/manual/diagrams.html) to be included in the vignette.

File: 07-contributor-guide.Rmd
Comment (Bai-Li-NOAA) [reply 07-contributor-guide.Rmd:178]: Yes, that's exactly what I had in mind. Removing all bulleted points below line 178 + last sentence of line 178.

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:112]: image
Confirmed that it does work.

File: 07-contributor-guide.Rmd
Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:41]: So how should we proceed? Should we leave the text for this section as is and turn your comment above into a new issue? Or, should we delete the text and turn your comment above into a new issue? Or, maybe some other option.

Comment (Bai-Li-NOAA) [reply 07-contributor-guide.Rmd:41]: Maybe keep the text for this section as is for now. I can create a new issue to explore moving the text into Markdown files and automatically generating a vignette from them. I can help with setting up the vignette and testing its functionality.

Comment (kellijohnson-NOAA) [reply 07-contributor-guide.Rmd:41]: Alright, sounds good.

**Comment by kellijohnson-NOAA** [comment]
Thanks @Bai-Li-NOAA for all of the helpful comments and Discussion. There are two tasks left for this PR,
- [ ] Find where the review guidance exists online for both sections and reference the links. This task is assiged to @kellijohnson-NOAA 
- [ ] Create a new issue about the adding README files to the FIMS folders that are talked about in 07-contributor-guide based on the discussion between @kellijohnson-NOAA and @Bai-Li-NOAA above. This task is assigned to @Bai-Li-NOAA 

The second task is not blocking the merge of this PR but the first is :). Once I check off and push the commit for task 1, then I will merge this into main.

**Comment by kellijohnson-NOAA** [comment]
The best documentation that I found for a review comes from [Google](https://google.github.io/eng-practices/review/reviewer/) and is already listed in the collaborative workflow. I did link to  [Conventional Comments](https://conventionalcomments.org/) and  the [Perforce Blog](https://www.perforce.com/blog/qac/9-best-practices-for-code-review) for some good practices of code reviews rather than having our own list.

This completes my tasks assigned to this PR, @Bai-Li-NOAA there is one additional task for you listed above but I will leave it to you to merge this PR whenever works best for you.

**Comment by Bai-Li-NOAA** [comment]
Great, I've created [a new issue](https://github.com/NOAA-FIMS/FIMS/issues/650) in the FIMS repo as discussed here.

**Pull request merged**
